keps/sig-api-machinery/20190820-master-service-of-type-externalname.md

```diff
@@ -0,0 +1,302 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The master kubernetes service created in the `default` namespace is called `kubernetes`. It's used by Pods running in the cluster to discover the kube-apiserver and communicate with it.
+
+kube-apiserver manages the master `Service` of type `ClusterIP` and associated `Endpoints` objects. Using this service, kubelet injects two environment variables called `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` into Pod's containers. Those environment variables can be detected by Kubernetes API clients such as `client-go` and used to determine if they run as a Pod in the cluster.
+
+This KEP is about adding support in Kubernetes for the master service to be of type `ExternalName` in addition to `ClusterIP`. This allows for kube-apiservers exposed via external FQDN (fully qualified domain name) to be directly consumed by Pods running in the cluster via that FQDN.
+
+## Motivation
+
+When exposing the kube-apiserver in HA mode, typically, a Layer-4 load-balancer is used. On different public/private cloud-providers, the endpoint of that load-balancer is either an IPv4/IPv6 address or a hostname (AWS being most prominent example for such offering). When consuming the latter offering, kubernetes cluster operators cannot use the hostname directly and have to disable the master endpoint reconciler in kube-apiserver with `--endpoint-reconciler-type=none` and write [custom controllers](https://github.com/gardener/aws-lb-readvertiser) which convert the hostname to IP addresses.
+
+It might be more suitable to advertise kube-apiserver's endpoint as FQDN, such as `api.my-cluster.example.com` for disaster recovery / prevention - in case of accidental deletion of said load-balancer and/or it's associated IP, the only thing required for recovery is to reconfigure the NS records to point to the newly created load-balancer.
+
+Historically there are several issues related to this problem:
+
+- https://github.com/kubernetes/kubernetes/issues/34043
+- https://github.com/kubernetes/kubernetes/issues/47588
```

Comment:
Maybe this comment still applies ?
https://github.com/kubernetes/kubernetes/pull/47588#issuecomment-316074996

> I think this has lots of ripples that haven't been considered:
> 
> interaction with version-skewed kubelets
> additional requirements for DNS names in the API server certs (if the ELB does TLS pass-through)
> breakage of x509 client cert auth (if the ELB does TLS termination)

Reply:
It does. 

> interaction with version-skewed kubelets

In the KEP, I've addressed lots of those problems. The safest way is to use featuregates.

> additional requirements for DNS names in the API server certs (if the ELB does TLS pass-through)

This is a requirement, but generation of certificates with those additional DNS names should not be a problem. [kubeadm](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/) already have the flag `--apiserver-cert-extra-sans` which can be used for adding such DNS names.

> breakage of x509 client cert auth (if the ELB does TLS termination)

Termination of TLS is a no-go in such cases. TCP or TLS-passthrough LBs/Proxies should be used.

Reply:
kubeadm is going out of tree, ETA 1.19, I think that depending on it is a bit risky

Reply:
I was using it as a reference for a tool offering such easy way of generation of such certificate and not as a reference. In fact this would be installation-specific.